lib/doorkeeper/models/access_token_mixin.rb

```diff
@@ -15,13 +15,14 @@ module AccessTokenMixin
         inverse_of: :access_tokens
       }
       if defined?(ActiveRecord::Base) && ActiveRecord::VERSION::MAJOR >= 5
-        belongs_to_options.merge(optional: true)
+        belongs_to_options.merge!(optional: true)
       end
 
       belongs_to :application, belongs_to_options
 
       validates :token, presence: true, uniqueness: true
       validates :refresh_token, uniqueness: true, if: :use_refresh_token?
+      validates :application_id, presence: true, allow_nil: true
```

Comment:
> I consider this a bit of a stretch, but it was intended to provide verbosity. Feel free to ditch it if needed. Validation based on the optional key still stands anyway.

I would ditch it indeed. I think it's more confusing than useful ("check this, but allow it to not be checked too").